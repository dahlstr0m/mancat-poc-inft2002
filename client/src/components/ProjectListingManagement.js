// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import { history } from '../index';
import { CardGrid, CardColumn, CardImage } from './Card';
import { Alert } from './Widgets';
import {
  projectService,
  posterService,
  type Project,
  type Poster,
} from '../services/portfolio-service';

/**
 * Renders project listing in management
 */
export default class ProjectListingManagement extends Component {
  projects: Project[] = [];
  posters: Poster[] = [];

  render() {
    return (
      <>
        <h1>Project management</h1>
        <CardGrid columns={1} columnsSm={2} columnsMd={3} columnsLg={4}>
          {this.projects
            .sort((projectA, projectB) => projectA.ranking - projectB.ranking) // Sorting to display by asc ranking
            .map((project) => (
              <CardColumn key={project.projectId}>
                <CardImage
                  button
                  img={this.getPosterUrl(project)}
                  title={project.title}
                  imgAlt={'Missing thumbnail for ' + project.title}
                  buttonText={'Manage project'}
                  buttonOnClick={() => history.push(`/admin/projects/${project.projectId}`)}
                >
                  Description: {project.projectDescription}
                  <br />
                  Status: {project.active == true ? 'Active' : 'Disabled'}
                </CardImage>
              </CardColumn>
            ))}
        </CardGrid>
      </>
    );
  }

  getPosterUrl(project: Project) {
    const poster = this.posters.find((poster) => poster.projectId === project.projectId);
    return poster ? poster.thumbnailUrl : '';
  }

  mounted() {
    projectService
      .getProjects()
      .then((projects) => (this.projects = projects))
      .catch((error: Error) => Alert.danger('Error getting projects: ' + error.message));

    posterService
      .getPosters()
      .then((posters) => (this.posters = posters))
      .catch((error: Error) => Alert.danger('Error getting posters: ' + error.message));
  }
}
