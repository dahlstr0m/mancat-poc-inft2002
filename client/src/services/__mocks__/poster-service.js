// @flow

/**
 * Mocked version of PosterService used in testing.
 */

type Poster = {
  posterId: number,
  projectId: number,
  posterDescription: string,
  posterUrl: string,
  thumbnailUrl: string,
};

class PosterService {
  getPosters() {
    return Promise.resolve([
      {
        posterId: 1,
        projectId: 1,
        posterDescription: 'Poster 1 description',
        posterUrl: 'Poster 1 url',
        thumbnailUrl: 'Poster 1 thumb url',
      },
      {
        posterId: 2,
        projectId: 2,
        posterDescription: 'Poster 2 description',
        posterUrl: 'Poster 2 url',
        thumbnailUrl: 'Poster 2 thumb url',
      },
      {
        posterId: 3,
        projectId: 3,
        posterDescription: 'Poster 3 description',
        posterUrl: 'Poster 3 url',
        thumbnailUrl: 'Poster 3 thumb url',
      },
    ]);
  }

  getPoster(posterId: number) {
    if (posterId === 1)
      return Promise.resolve({
        posterId: 1,
        projectId: 1,
        posterDescription: 'Poster 1 description',
        posterUrl: 'Poster 1 url',
        thumbnailUrl: 'Poster 1 thumb url',
      });

    if (posterId === 2)
      return Promise.resolve({
        posterId: 2,
        projectId: 2,
        posterDescription: 'Poster 2 description',
        posterUrl: 'Poster 2 url',
        thumbnailUrl: 'Poster 2 thumb url',
      });
  }

  getProjectPosters(projectId: number) {
    if (projectId === 1)
      return Promise.resolve({
        posterId: 1,
        projectId: 1,
        posterDescription: 'Poster 1 description',
        posterUrl: 'Poster 1 url',
        thumbnailUrl: 'Poster 1 thumb url',
      });

    if (projectId === 2)
      return Promise.resolve({
        posterId: 2,
        projectId: 2,
        posterDescription: 'Poster 2 description',
        posterUrl: 'Poster 2 url',
        thumbnailUrl: 'Poster 2 thumb url',
      });

    if (projectId === 3)
      return Promise.resolve({
        posterId: 3,
        projectId: 3,
        posterDescription: 'Poster 3 description',
        posterUrl: 'Poster 3 url',
        thumbnailUrl: 'Poster 3 thumb url',
      });
  }

  createPoster(poster: Poster) {
    return Promise.resolve(4);
  }

  updatePoster(poster: Poster) {
    return Promise.resolve();
  }

  deletePoster(posterId: number) {
    return Promise.resolve();
  }
}

const posterService = new PosterService();
export default posterService;
