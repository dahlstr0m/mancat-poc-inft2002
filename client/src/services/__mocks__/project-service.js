// @flow

type Project = {
  projectId: number,
  title: string,
  projectDescription: string,
  projectDate: string,
  categoryId: number,
  employerId: number,
  ranking: number,
  active: boolean,
};

class ProjectService {
  getProjects() {
    return Promise.resolve([
      {
        projectId: 1,
        title: 'Project 1',
        projectDescription: 'Project 1 description',
        projectDate: '10-10-2020',
        categoryId: 1,
        employerId: 1,
        ranking: 1,
        active: true,
      },
      {
        projectId: 2,
        title: 'Project 2',
        projectDescription: 'Project 2 description',
        projectDate: '11-10-2020',
        categoryId: 2,
        employerId: 2,
        ranking: 2,
        active: true,
      },
      {
        projectId: 3,
        title: 'Project 3',
        projectDescription: 'Project 3 description',
        projectDate: '12-10-2020',
        categoryId: 3,
        employerId: 3,
        ranking: 3,
        active: true,
      },
    ]);
  }

  getProject(projectId: number) {
    if (projectId === 1)
      return Promise.resolve({
        projectId: 1,
        title: 'Project 1',
        projectDescription: 'Project 1 description',
        projectDate: '10-10-2020',
        categoryId: 1,
        employerId: 1,
        ranking: 1,
        active: true,
      });

    if (projectId === 2)
      return Promise.resolve({
        projectId: 2,
        title: 'Project 2',
        projectDescription: 'Project 2 description',
        projectDate: '11-10-2020',
        categoryId: 2,
        employerId: 2,
        ranking: 2,
        active: true,
      });
  }

  createProject(project: Project) {
    return Promise.resolve(4);
  }

  updateProject(project: Project) {
    return Promise.resolve();
  }

  deleteProject(projectId: number) {
    return Promise.resolve();
  }
}

const projectService = new ProjectService();
export default projectService;
