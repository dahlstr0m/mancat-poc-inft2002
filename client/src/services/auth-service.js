// @flow

import axios from 'axios';

axios.defaults.baseURL = 'http://localhost:3000/api/v1';

export type User = {
  username: string,
  password: string,
};

class AuthService {
  authenticated: boolean;

  constructor() {
    this.authenticated = false;
  }

  login(user: User) {
    return axios
      .post<User, {}>('/auth/login', user)
      .then(() => (this.authenticated = true))
      .catch((error) => console.log(error));
  }

  isAuthenticated() {
    return this.authenticated;
  }
}

const authService = new AuthService();
export default authService;
