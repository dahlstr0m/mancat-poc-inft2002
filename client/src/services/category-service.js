// @flow
import axios from 'axios';

axios.defaults.baseURL = 'http://localhost:3000/api/v1';

export type Category = {
  categoryId: number,
  categoryName: string,
};

class CategoryService {
  getCategories() {
    return axios.get<Category[]>('/categories').then((response) => response.data);
  }

  getCategory(categoryId: number) {
    return axios.get<Category>('/categories/' + categoryId).then((response) => response.data);
  }

  getProjectCategory(projectId: number) {
    return axios.get<Category>(`/projects/${projectId}/category`).then((response) => response.data);
  }

  createCategory(category: Category) {
    return axios
      .post<Category, { categoryId: number }>('/categories', category)
      .then((response) => response.data);
  }

  updateCategory(category: Category) {
    return axios
      .put<Category, void>(`/categories/${category.categoryId}`, category)
      .then((response) => response.data);
  }

  deleteCategory(categoryId: number) {
    return axios.delete<void>('/categories/' + categoryId).then((response) => response.data);
  }
}

const categoryService = new CategoryService();
export default categoryService;
