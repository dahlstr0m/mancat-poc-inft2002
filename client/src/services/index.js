export * from './auth-service';
export * from './portfolio-service';
