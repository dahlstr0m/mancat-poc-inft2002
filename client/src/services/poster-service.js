// @flow
import axios from 'axios';

axios.defaults.baseURL = 'http://localhost:3000/api/v1';

export type Poster = {
  posterId: number,
  projectId: number,
  posterDescription: string,
  posterUrl: string,
  thumbnailUrl: string,
};

class PosterService {
  getPosters() {
    return axios.get<Poster[]>('/posters').then((response) => response.data);
  }

  getPoster(posterId: number) {
    return axios.get<Poster>('/posters/' + posterId).then((response) => response.data);
  }

  getProjectPosters(projectId: number) {
    return axios.get<Poster[]>(`/projects/${projectId}/posters`).then((response) => response.data);
  }

  createPoster(poster: Poster) {
    return axios
      .post<Poster, { posterId: number }>('/posters', poster)
      .then((response) => response.data);
  }

  updatePoster(poster: Poster) {
    return axios
      .put<Poster, void>(`/posters/${poster.posterId}`, poster)
      .then((response) => response.data);
  }

  deletePoster(posterId: number) {
    return axios.delete<void>('/posters/' + posterId).then((response) => response.data);
  }
}

const posterService = new PosterService();
export default posterService;
