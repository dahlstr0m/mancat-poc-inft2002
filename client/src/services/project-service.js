// @flow
import axios from 'axios';

axios.defaults.baseURL = 'http://localhost:3000/api/v1';

export type Project = {
  projectId: number,
  title: string,
  projectDescription: string,
  projectDate: string,
  categoryId: number,
  employerId: number,
  ranking: number,
  active: boolean,
};

class ProjectService {
  getProjects() {
    return axios.get<Project[]>('/projects').then((response) => response.data);
  }

  getProject(projectId: number) {
    return axios.get<Project>('/projects/' + projectId).then((response) => response.data);
  }

  createProject(project: Project) {
    return axios
      .post<Project, { projectId: number }>('/projects', project)
      .then((response) => response.data);
  }

  updateProject(project: Project) {
    return axios
      .put<Project, void>(`/projects/${project.projectId}`, project)
      .then((response) => response.data);
  }

  /**
   * Special method to update every projects ranking at once.
   * Used only in <ProjectRanking />.
   */
  updateRanking(projects: Project[]) {
    return axios
      .put<Project[], void>('/projects/ranking', projects)
      .then((response) => response.data);
  }

  deleteProject(projectId: number) {
    return axios.delete<void>('/projects/' + projectId).then((response) => response.data);
  }
}

const projectService = new ProjectService();
export default projectService;
