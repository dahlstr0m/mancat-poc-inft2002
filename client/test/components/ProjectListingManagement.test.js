// @flow

import * as React from 'react';
import { shallow } from 'enzyme';

// Types
import { type Project } from '../../src/services/project-service';
import { type Poster } from '../../src/services/poster-service';

// Components
import ProjectListingManagement from '../../src/components/ProjectListingManagement';

// Mocks
jest.mock('../../src/services/project-service');
jest.mock('../../src/services/poster-service');

// Tester
describe('ProjectListingManagement component tests', () => {
  test('ProjectListingManagement draws correctly', (done) => {
    const wrapper = shallow(<ProjectListingManagement />);

    // Wait for events to complete
    setTimeout(() => {
      expect(wrapper).toMatchSnapshot();

      done();
    });
  });
});
