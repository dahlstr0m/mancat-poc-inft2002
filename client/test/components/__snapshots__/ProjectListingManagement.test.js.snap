// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectListingManagement component tests ProjectListingManagement draws correctly 1`] = `
<Fragment>
  <h1>
    Project management
  </h1>
  <CardGrid
    columns={1}
    columnsLg={4}
    columnsMd={3}
    columnsSm={2}
  >
    <CardColumn
      key="1"
    >
      <CardImage
        button={true}
        buttonOnClick={[Function]}
        buttonText="Manage project"
        img="Poster 1 thumb url"
        imgAlt="Missing thumbnail for Project 1"
        title="Project 1"
      >
        Description: 
        Project 1 description
        <br />
        Status: 
        Active
      </CardImage>
    </CardColumn>
    <CardColumn
      key="2"
    >
      <CardImage
        button={true}
        buttonOnClick={[Function]}
        buttonText="Manage project"
        img="Poster 2 thumb url"
        imgAlt="Missing thumbnail for Project 2"
        title="Project 2"
      >
        Description: 
        Project 2 description
        <br />
        Status: 
        Active
      </CardImage>
    </CardColumn>
    <CardColumn
      key="3"
    >
      <CardImage
        button={true}
        buttonOnClick={[Function]}
        buttonText="Manage project"
        img="Poster 3 thumb url"
        imgAlt="Missing thumbnail for Project 3"
        title="Project 3"
      >
        Description: 
        Project 3 description
        <br />
        Status: 
        Active
      </CardImage>
    </CardColumn>
  </CardGrid>
</Fragment>
`;
